crates/ruff/src/jupyter/index.rs

```diff
@@ -0,0 +1,96 @@
+use std::iter;
+use std::sync::Arc;
+
+use ruff_newlines::NewlineWithTrailingNewline;
+
+use crate::jupyter::{Cell, SourceValue};
+
+/// Jupyter Notebook indexing table
+///
+/// When we lint a jupyter notebook, we have to translate the row/column based on
+/// [`ruff_text_size::TextSize`] to jupyter notebook cell/row/column.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct JupyterIndex {
+    pub(super) inner: Arc<JupyterIndexInner>,
+}
+
+impl JupyterIndex {
+    /// Returns the cell number (1-based) for the given row (1-based).
+    pub fn get_cell(&self, row: usize) -> u32 {
+        self.inner.row_to_cell[row]
```

Comment:
So, these are only used when emitting the diagnostics. How do you think we should handle the `None` case?

If we use the default value, that would render the diagnostic at the top of file.

This is not similar but I think `source_location` panics as well if it's out of bounds.

Reply:
> This is not similar but I think source_location panics as well if it's out of bounds.

That's true and we should probably avoid that too... It's unfortunate if ruff dies just because it fails to render a single diagnostic (This isn't meant that YOU should change it, it should probably be me)

> So, these are only used when emitting the diagnostics. How do you think we should handle the None case?

I think that's' a reasonable fallback and provides better usability than taking down ruff. I would expect users to report a bug that we can then fix.